### PR TITLE
Quickstart 43

### PR DESCRIPTION
**Pull request by baranowb**
As in subject. Wasnt aware that hibernatedemo has been completed so I hacked around basic numberguess.

### CODE REVIEW

**Comment by pmuir** [comment]
Once you've applied the same changes as I suggested for 44 to this quickstart, just post a comment and I will review.

**Comment by baranowb** [comment]
Updated, with some small tweaks Ive found.

**Comment by sgilda** [comment]
Pete, I'm not sure how to get to the comments for the closed pull request for the log4j demo, so I'm not sure what changes you requested. The README looks good other than a minor typo in this sentence "Even thought this examples java code is simple, it is quite complicated example." It should probably read "Even though this example's java code is simple, it is a complicated example", but I don't think it is enough to hold it up.
Looks good. Nice example!

File: hibernate3/pom.xml [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		<!-- Define hibernate-commons-annotations version -->
+		<hibernate.commons.annotations.version>3.2.0.Final</hibernate.commons.annotations.version>
+	</properties>
+
+	<!-- NOTE: -->
+	<dependencyManagement>
+		<dependencies>
+			<!--  Define hiberante artifacts versions  -->
+			<!-- Define dependency version of hibernate-core 
+			    NOTE! that scope is	not "provided", this will make maven.war.plugin to include it in WEB-INF/lib -->
```
Comment (pmuir) [inline hibernate3/pom.xml]: I would call it the "Maven war plugin" without the periods.

File: hibernate3/pom.xml [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+			<dependency>
+				<groupId>org.hibernate</groupId>
+				<artifactId>hibernate-core</artifactId>
+				<version>${hibernate.core.version}</version>
+				
+				<!-- Define exclusion of artifacts. Those artifacts have scope set to 'compile', to avoid
```
Comment (pmuir) [inline hibernate3/pom.xml]: I would this as "excluding some of the transitive dependencies of Hibernate 3, as we include them as modules from JBoss AS".

File: hibernate3/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	* when container has module with different version of binary dependency</li>
+
+This example falls into third category. It shows how to include and configure Hibernate 3.x.
+To better understand how it is done, its good to get acquainted with 'log4jdemo' quickstart.
+
+ Even thought this examples java code is simple, it is quite complicated example. As shown in 'log4jdemo' defining dependency is quite easy.
+ This example, however goes a bit beyond simple module dependency.<br>
+ There are three aspects of this example that play a role here:
+
+ 	* assembly - the Maven script and maven-war-plugin create 'war' which includes ONLY Hibernate 3.x binaries. To understand better how its
```
Comment (pmuir) [inline hibernate3/README.md]: This is a very confusing thing to call it, as "assembly" in maven usually refers to the maven assembly plugin. Instead, I would call it war archive creation and libraries, or something like that...

**Comment by pmuir** [comment]
Regardless of my comments above, I really think we should have hibernate4 and 3 demonstrated using exactly the same code _except_ for the critical differences, so I think we need to redo this quickstart to be built on top of the hibernate4 demo (or vs versa). I'll move this to email.

**Comment by sgilda** [comment]
I agree. It's easier for a user to compare the 2 different versions of hibernate if they are doing the same thing.

**Comment by baranowb** [comment]
Switch to h4 example source is quite straightforward, since most of the problems faced here are in conf/binary deps.
Did not do that since when I've started to hack this quickstart, h4 quickstart was not there.

**Comment by baranowb** [comment]
Pete please have a look, hope I did not forget to update according to comments. Cheers.

**Comment by sgilda** [comment]
--- I would like to suggest a change to the "What is it?" section of the README file to make it easier to read. Maybe something like the following:

This is a simple JSF 2.0 example based on _hibernate4-demo_ quickstart.  Its purpose is to demonstrate how you can use Hibernate 3 in your application.

This quickstart, like _log4jdemo_, demonstrates how to define a module dependency. However, this quickstart goes beyond that and also demonstrates the following:

```
* WAR creation - The Maven script and Maven WAR plugin create a *WAR* archive that includes ONLY the Hibernate 3.x binaries. To understand better how this is achieved, please refer to the "pom.xml" in the root directory of this quickstart. Additional information can be found in the <http://maven.apache.org/plugins/maven-war-plugin> documentation.
* Module exclusion and inclusion - This example demonstrates how to control class loading using "dependencies" and "exclusions" in the *jboss-deployment-structure.xml* file. For more information about this file, please refer to <https://docs.jboss.org/author/display/AS7/Developer+Guide#DeveloperGuide-JBossDeploymentStructureFile>
* Persistence configuration - Configuration is required to tell the container how to load JPA/Hibernate.
```

--- We can keep the URL "hibernate3" or modify it to "hibernate3-demo". Either way is fine by me as long as the hibernate 4 demo is named in a similar manner.

**Comment by sgilda** [comment]
Since the hibernatedemo for Hibernate 4 is being renamed to hibernate4-demo, could we rename this one to hibernate3-demo so they are in sync?

**Comment by pmuir** [comment]
Actually I would go the other way. We don't include the -demo suffix on anything else, so why should we here. Just call it hibernate3 and hibernate4!

**Comment by sgilda** [comment]
I'm fine either way. As long as they both have the hibernate release.

**Comment by sgilda** [comment]
This looks great! I've reviewed this for documentation and comments and it looks good. I also tested the deployment and runtime and it works fine.
A couple of comments:
-- This example needs a unique table name specified in the Member.java and import.sql file to avoid a conflict with the other quickstart that already uses "Member". Maybe "MemberHibernate3"?
-- The URL for this quickstart is: http://localhost:8080/jboss-as-hibernate3/
    The URL for the hibernate4 quickstart is: http://localhost:8080/hibernate4/
    It would be nice if they were consistent. Since most of the URLs start with "jboss-as-", maybe the URL for the 
    hibernate4 should change? Pete, do you have a preference?
--This is just a suggestion:  You've commented the code differences between Hibernate 3 and Hibernate 4 in the classes, but it might be nice to give an overview of the differences the README file. Tell them to look for _NOTE_ comments and let them know which files or classes they need to look at to find the changes and compare with the hibernate4 quickstart.

Pete, do you need to do a code review on this one?

File: hibernate3/pom.xml [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+			NOTE: this is a grey area, changing those versions can cause failure. 
+			NOTE2: hibernate-annotations are not included here, since hibernata-core 3.6.X contains former artifact. -->
+		<!-- Define hibernate-core version -->	
+		<hibernate.core.version>3.6.8.Final</hibernate.core.version>
+		<!-- Define hibernate-entitymanager version -->
+		<hibernate.entitymanager.version>3.6.8.Final</hibernate.entitymanager.version>
+		<!-- Define hibernate-validator version -->
+		<hibernate.validator.version>3.1.0.GA</hibernate.validator.version>
+		<!-- Define hibernate-commons-annotations version -->
+		<hibernate.commons.annotations.version>3.2.0.Final</hibernate.commons.annotations.version>
```
Comment (pmuir) [inline hibernate3/pom.xml]: Why don't we move the version numbers out of properties and into depMgmt, that will make this simpler to comprehend.

File: hibernate3/pom.xml [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	</properties>
+
+	<!-- NOTE: -->
+	<dependencyManagement>
+		<dependencies>
+			<!--  Define hiberante artifacts versions  -->
+			<!-- Define dependency version of hibernate-core 
+			    NOTE! that scope is	not "provided", this will make Maven war plugin to include it in WEB-INF/lib -->
```
Comment (pmuir) [inline hibernate3/pom.xml]: I would invert this negative - "scope is "compile"..."

File: hibernate3/pom.xml [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+					<!-- Exclude jta artifact -->
+					<exclusion>
+						<groupId>javax.transaction</groupId>
+						<artifactId>jta</artifactId>
+					</exclusion>
+				</exclusions>
+			</dependency>
+
+			<!-- Define dependency version of hibernate-entitymanager: NOTE! that scope is 
+				not "provided", this will make Maven war plugin to include it in WEB-INF/lib -->
```
Comment (pmuir) [inline hibernate3/pom.xml]: Again, I would invert this negative.

File: hibernate3/pom.xml [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+			<dependency>
+				<groupId>org.hibernate</groupId>
+				<artifactId>hibernate-entitymanager</artifactId>
+				<version>${hibernate.entitymanager.version}</version>
+				<!-- Excluding some of the transitive dependencies of Hibernate 3, as we include them as modules from JBoss AS7 -->
```
Comment (pmuir) [inline hibernate3/pom.xml]: I would switch this sentence around "Some transitive dependencies of Hibernate 3 are available in JBoss AS 7 as modules, so we don't include them in WEB-INF/lib, but instead depend on the modules".

File: hibernate3/pom.xml [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+			<!-- Define dependency version of hibernate-core 
+			    NOTE! that scope is	not "provided", this will make Maven war plugin to include it in WEB-INF/lib -->
+			<dependency>
+				<groupId>org.hibernate</groupId>
+				<artifactId>hibernate-core</artifactId>
+				<version>${hibernate.core.version}</version>
+				
+				<!-- Excluding some of the transitive dependencies of Hibernate 3, as we include them as modules from JBoss AS7 -->
+				<exclusions>
+					<!-- Exclude antlr artifact -->
```
Comment (pmuir) [inline hibernate3/pom.xml]: I'm not sure comments like this are really helpful, they aren't adding any information to the user.

File: hibernate3/pom.xml [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		<hibernate.core.version>3.6.8.Final</hibernate.core.version>
+		<!-- Define hibernate-entitymanager version -->
+		<hibernate.entitymanager.version>3.6.8.Final</hibernate.entitymanager.version>
+		<!-- Define hibernate-validator version -->
+		<hibernate.validator.version>3.1.0.GA</hibernate.validator.version>
+		<!-- Define hibernate-commons-annotations version -->
+		<hibernate.commons.annotations.version>3.2.0.Final</hibernate.commons.annotations.version>
+	</properties>
+
+	<!-- NOTE: -->
```
Comment (pmuir) [inline hibernate3/pom.xml]: This comment appears unused.

File: hibernate3/pom.xml [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+	<dependencyManagement>
```
Comment (pmuir) [inline hibernate3/pom.xml:32]: We always place the scope in <dependency> not <dependencyManagement> and we should do the same here.

File: hibernate3/pom.xml [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+         in JBoss AS 7 -->
+      <dependency>
+         <groupId>org.jboss.spec.javax.ejb</groupId>
+         <artifactId>jboss-ejb-api_3.1_spec</artifactId>
+         <scope>provided</scope>
+      </dependency>
+
+
+
+		<!-- Define dependency on hibernate 3.x, since scope is 'compile', it will
```
Comment (pmuir) [inline hibernate3/pom.xml:192]: The whitespace is wrong here.

File: hibernate3/README.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+========================
+
+What is it?
+-----------
+
+This is a simple JSF 2.0 example based on hibernate4 quickstart. Its purpose is to demonstrate how you can use Hibernate 3 in your application.
+
+This quickstart, like log4j quickstart , demonstrates how to define a module dependency. However, this quickstart goes beyond that and also demonstrates the following:
+ 
+ 	* WAR creation - The Maven script and Maven WAR plugin create a *WAR* archive that includes ONLY the Hibernate 3.x binaries. To understand better how this is achieved, please refer to the *pom.xml* in the root directory of this quickstart. Additional information can be found in the <http://maven.apache.org/plugins/maven-war-plugin> documentation.
```
Comment (pmuir) [inline hibernate3/README.md]: Whitespace is wrong. Not sure what you are doing specially with the war plugin here?

File: hibernate3/src/main/java/org/jboss/as/quickstart/hibernate3/model/Member.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+   @Pattern(regex = "[A-Za-z ]*", message = "must contain only letters and spaces")
+   private String name;
+
+   @NotNull
+   @NotEmpty
+   @Email
+   private String email;
+
+   @NotNull
+   //NOTE: @Size in v3 is used only for arrays and collection objects, we have to use @Length for string validation
```
Comment (pmuir) [inline hibernate3/src/main/java/org/jboss/as/quickstart/hibernate3/model/Member.java]: Refer to to it as Hibernate 3, not v3. I've never heard it called v3 before!

**Comment by pmuir** [comment]
My comments are in.

**Comment by sgilda** [comment]
Waiting for changes requested by Pete to be implemented.

**Comment by baranowb** [comment]
Sorry, Im currently testing a release, should be done by tomorrow.

**Comment by sgilda** [comment]
No problem. The comment was mainly for my own benefit so I would know where I stand on this one. :-)

File: hibernate3/pom.xml (hunk 10 above)
Comment (baranowb) [inline hibernate3/pom.xml:32]: Pete Im not able to make sense off this comment :) could you please explain?

Comment (pmuir) [inline hibernate3/pom.xml:32]: Sorry, the scope is placed in `<dependency>` not `<dependencyManagement>` and you should put your comments in there about scope...

File: hibernate3/README.md (hunk 12 above)
Comment (baranowb) [inline hibernate3/README.md]: Removed, as on second thought it seemed not important, since its implied by quickstart purpose.

**Comment by baranowb** [comment]
Incorporated all comments I think.

**Comment by sgilda** [comment]
This looks great! The only change needed at this point is the version for the jboss-as-maven-plugin artifact in pom.xml file should be 7.1.0.CR1 instead of 7.1.0.Beta1b. This is my fault because it took too long for me to review it and it changed in the meantime.  Current versions can be found here: https://docs.jboss.org/author/display/AS71/Maven+POM+Versions+Checklist
I have one minor comment: Why does the address label in the page have a # at the end, like this "Address #:" when the seeded data is City, State? :-)

**Comment by sgilda** [comment]
Sorry. One more comment. Member.java is confusing to me. You have:
// import validator annotations, those are Hiberante3, check this class source code and commented imports
// for better understanding of differences.
followed by some imports. There's a typo in Hibernate3 above.
Then 
//NOTE: those are Hiberante4 imports and dependencies ie. validation-api. Hibernate3 does not
//import org.hibernate.validator.constraints.Email;
//import org.hibernate.validator.constraints.NotEmpty;
//import javax.validation.constraints.Digits;
//import javax.validation.constraints.NotNull;
//import javax.validation.constraints.Pattern;
//import javax.validation.constraints.Size;
Are you trying to show the difference between what you import for Hibernate version 3 versus what you import for Hibernate 4? 
If that's the case, I'm not sure it belongs here because it clutters up the code. it might be clearer to point out the differences in the imports, package names, and classes for Hibernate3 in the README file.
Pete, what do you think?

**Comment by pmuir** [comment]
Sande, agreed, it's better to not have commented out code in an example. We have the hibernate4 example for people to do a side-by-side comparison, should they wish. We can call out prominent differences, such as this in the README.md.

**Comment by baranowb** [comment]
Well I actually disagree. If ppl have two screens than its easy to see he differences without having them pointed out like that.
However Ive removed comments, did not add them in readme since its not a place for it. Changed plugin version.

**Comment by sgilda** [comment]
Hi Bartosz,
I agree the README does not need that level of detail. What I had in mind was something like this:

If you compare the source code in the hibernate4 to this example, you will see the following differences:
- Some of the Class package names have changed with Hibernate 4
- Hibernate 3 used @Length for String validation while Hibernate 4 uses @Size 
- In Hibernate 3, the term for regular expression in the Pattern annotation was 'regex'. In Hibernate 4 the term is 'regexp'
- The terms within the Digits annotation have also changed.

But, it certainly isn't necessary. This looks good to me. 
Pete, did you have any other comments? If not I think this one is ready.

**Comment by sgilda** [comment]
I attempted a test push to my own repo and had a lot of issues and conflicts. Waiting for advice from Pete as to what I did wrong. This was the output:
quickstart (hibernate3) $ git rebase upstream/master
First, rewinding head to replay your work on top of it...
Applying: quickstart-43
Applying: quickstart-43
Applying: small update
Applying: As discussed: switch to hibernatedemo source so both quickstarts are alike
Applying: As discussed: switch to hibernatedemo source so both quickstarts are alike
Using index info to reconstruct a base tree...
<stdin>:75: trailing whitespace, space before tab in indent.
     <li>war creation</li> - the Maven script and maven-war-plugin create 'war' which includes ONLY Hibernate 3.x binaries. To understand better how its
<stdin>:109: trailing whitespace.

<stdin>:110: trailing whitespace, space before tab in indent.
     \* war creation - the Maven script and maven war plugin create _war_ which includes ONLY Hibernate 3.x binaries. To understand better how its
<stdin>:111: space before tab in indent.
         achieved, please refer to this examples 'pom.xml' and http://maven.apache.org/plugins/maven-war-plugin documentation.
<stdin>:115: space before tab in indent.
         in application deployment unit. It also excludes implicit dependency on container default Hibernate module (which is version 4.x).
warning: squelched 33 whitespace errors
warning: 38 lines add whitespace errors.
Falling back to patching base and 3-way merge...
CONFLICT (rename/delete): Rename hibernate3/src/main/java/org/jboss/as/quickstarts/hibernate3/Game.java->numberguess/src/main/java/org/jboss/as/quickstarts/numberguess/Game.java in HEAD and deleted in As discussed: switch to hibernatedemo source so both quickstarts are alike
CONFLICT (rename/delete): Rename hibernate3/src/main/java/org/jboss/as/quickstarts/hibernate3/Generator.java->numberguess/src/main/java/org/jboss/as/quickstarts/numberguess/Generator.java in HEAD and deleted in As discussed: switch to hibernatedemo source so both quickstarts are alike
CONFLICT (rename/delete): Rename hibernate3/src/main/java/org/jboss/as/quickstarts/hibernate3/MaxNumber.java->numberguess/src/main/java/org/jboss/as/quickstarts/numberguess/MaxNumber.java in HEAD and deleted in As discussed: switch to hibernatedemo source so both quickstarts are alike
CONFLICT (rename/delete): Rename hibernate3/src/main/java/org/jboss/as/quickstarts/hibernate3/Random.java->numberguess/src/main/java/org/jboss/as/quickstarts/numberguess/Random.java in HEAD and deleted in As discussed: switch to hibernatedemo source so both quickstarts are alike
CONFLICT (rename/delete): Rename hibernate3/src/main/webapp/home.xhtml->numberguess/src/main/webapp/home.xhtml in HEAD and deleted in As discussed: switch to hibernatedemo source so both quickstarts are alike
Failed to merge in the changes.
Patch failed at 0005 As discussed: switch to hibernatedemo source so both quickstarts are alike

**Comment by baranowb** [comment]
Im affraid Im as green as you are on this, this is atually the first time Im using git.

**Comment by pmuir** [comment]
Here is a rebased version of the pull https://github.com/jbossas/quickstart/pull/79. I'll close this one in favor of that one.

**Comment by pmuir** [comment]
Please review my pull, make sure it matches what you think it should do.

**Pull request closed without merging**